server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryServer.java

```diff
@@ -243,7 +246,12 @@ public void onNext(DiscoveryRequest request) {
           nonce,
           request.getVersionInfo());
 
-      callbacks.forEach(cb -> cb.onStreamRequest(streamId, request));
+      try {
+        callbacks.forEach(cb -> cb.onStreamRequest(streamId, request));
+      } catch (RequestException e) {
+        closeWithError(e);
```

Comment:
Is the `closeWithError` method with the `AtomicBoolean` logic needed here? I think this could just directly call `responseObserver.onError` like the other error condition above.

Reply:
I think it is needed. In my manual testing, without that logic I encountered following error 
```
SEVERE: Exception while executing runnable io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1MessagesAvailable@1436e222: 
java.lang.IllegalStateException: call already closed
```

It is not deterministic. I tried to reproduce it again today many times without success. And then something "switches" and right now I receive this error every time I run my tests.

It seems that it happens when Envoy sends two DiscoveryRequest simultanously on ADS stream - one for `type.googleapis.com/envoy.api.v2.Listener[]` and one for `type.googleapis.com/envoy.api.v2.Cluster[]`. In my manual tests it didn't happen until third stream is opened. On first two streams Envoy asked only for Clusters (in my test every stream is closed by throwing `RequestException` in `DiscoveryServerCallbacks`). Maybe it has something to do with Envoy initialization process, because third stream is opened after Envoy log message `all dependencies initialized.`

About other places where we call `responseObserver.onError` - I think it is possible that places should check against `isClosing` flag too.

That error is not fatal - it seems that its only consequence is log message. But I think we don't want error logs with level `SEVERE` as part of normal application flow.

Reply:
Sorry about the delay. Makes sense. I'll add a follow-up issue to look into a more general solution to the problem, but I think the current approach is good for now.